### PR TITLE
Add optimization barrier to enable the gradient checkpointing

### PR DESCRIPTION
**Pull request by JackCaoG**
This is to fix https://github.com/pytorch/xla/issues/3455. 

This pr introduced a new api `xm.optimization_barrier`, the usage is pretty straightforward, you call this api on the output of the forward pass. For example

```
        dummy_data = torch.zeros(64, 1024, 14, 14, device=device)
        optimizer.zero_grad()
        x = dummy_data
        for n_l, layer in enumerate(model):
            if n_l > 0 and grad_checkpoint:
                x = torch.utils.checkpoint.checkpoint(layer, x)
            else:
                x = layer(x)

        (x,) = xm.optimization_barrier([x])

        dummy_loss = x.sum()
        dummy_loss.backward()
        optimizer.step()
```

this ensures that all of its inputs(in this case the result of the forward pass) are evaluated before any operators that depend on the `optimization_barrier`'s outputs, and XLA is forbidden from eliding or moving non-trivial computations across `optimization_barrier` operators.

running a mini repo on my end, without `optimization_barrier`, I see
```
step 198, free memory after forward = 11514304
step 198, free memory = 11514304
step 199, free memory after forward = 11514304
step 199, free memory = 11514304
Metric: ExecuteTime
  TotalSamples: 200
  Accumulator: 41s329ms857.374us
  ValueRate: 590ms547.571us / second
  Rate: 2.85296 / second
  Percentiles: 1%=204ms379.962us; 5%=205ms526.820us; 10%=205ms608.515us; 20%=205ms785.190us; 50%=205ms188.236us; 80%=206ms065.475us; 90%=207ms667.218us; 95%=218ms647.765us; 99%=259ms730.178us
```

with `optimization barrier` I saw
```
step 198, free memory after forward = 15487456
step 198, free memory = 15487456
step 199, free memory after forward = 15487456
step 199, free memory = 15487456
Metric: ExecuteTime
  TotalSamples: 200
  Accumulator: 12s261ms783.098us
  ValueRate: 763ms711.836us / second
  Rate: 12.4415 / second
  Percentiles: 1%=061ms649.505us; 5%=061ms756.333us; 10%=061ms832.865us; 20%=061ms923.435us; 50%=061ms077.979us; 80%=061ms220.387us; 90%=061ms284.215us; 95%=061ms348.149us; 99%=081ms290.276us
```

What is a bit puzzling is that execute time improved. I would expect execute time to increase since cse was not able to collapse the two forward pass computation. It is a nice surprise so I will take it.

fyi @ronghanghu

### CODE REVIEW

**Comment by JackCaoG** [comment]
It is a bit tricky to test op as it does not work on XLA:CPU, I want to unblock user first so I might merged this pr with only manual testing.

**Comment by ronghanghu** [comment]
Thanks for the new API! A question regarding the usage

Suppose we have a model with a forward pass that takes inputs a, b, and c, and returns two variables x and y
```
x, y = torch.utils.checkpoint.checkpoint(layer, a, b, c)
```

To use this API and avoid fusing x and y with the backward pass, should I separately run
```
x = xm.optimization_barrier(x)
y = xm.optimization_barrier(y)
```
before my backward pass, or do I only need to call this API on one of x and y, or should I use it in another way?

**Comment by JackCaoG** [comment]
You should do
```
x = xm.optimization_barrier(x)
y = xm.optimization_barrier(y)
```

That being said, It might be a good idea for me to make this api to accept a list of tensor instead of a single tensor.

**Comment by ronghanghu** [comment]
I guess either interface is okay. If it takes a single tensor, I can also call it in one line like
```
x, y = map(xm.optimization_barrier, (x, y))
```
and apply similarly it to other cases (e.g. when the model output is a dictionary)

**Comment by JackCaoG** [comment]
manually verified that test passed on TPU, will merge this today(after build passed)

**Comment by JackCaoG** [comment]
failure is related to https://github.com/pytorch/xla/commit/3e3255e413ef01516fbc7112a8e53882ac0fb5f7 not being included, merge the pr to unblock user.
